rskj-core/src/main/java/org/ethereum/db/ReceiptStoreImplV2.java

```diff
@@ -0,0 +1,134 @@
+        } else {
+            // save tx receipt data as a separate item first
+            byte[] key = getCombinedKey(txHash, blockHash);
+            TransactionInfo newTxInfo = new TransactionInfo(receipt, blockHash, transactionIndex);
+            receiptsDS.put(key, newTxInfo.getEncoded());
+
+            // add block hash to the list
+            byte[][] blockHashArr = new byte[txListSize + 1][];
+            for (int i = 0; i < txListSize; ++i) {
+                blockHashArr[i] = RLP.encodeElement(txList.get(i).getRLPData());
+            }
+
+            blockHashArr[txListSize] = RLP.encodeElement(blockHash);
+
+            receiptsDS.put(txHash, RLP.encodeList(blockHashArr));
+        }
+    }
+
+    @Override
+    public Optional<TransactionInfo> get(byte[] transactionHash, byte[] blockHash) {
+        // try first a new data format
+        Optional<TransactionInfo> txInfoOpt = getInternal(transactionHash, blockHash);
+        if (txInfoOpt.isPresent()) {
+            return txInfoOpt;
+        }
+
+        // fallback with older data format
+        return super.get(transactionHash, blockHash);
+    }
+
+    @Override
+    public Optional<TransactionInfo> getInMainChain(byte[] transactionHash, BlockStore store) {
+        // try first a new data format
+        byte[] txInfoBytes = receiptsDS.get(transactionHash);
+        List<byte[]> blockHashList = parseBlockHashList(txInfoBytes);
+        if (!blockHashList.isEmpty()) {
+            for (byte[] blockHash : blockHashList) {
+                Block block = store.getBlockByHash(blockHash);
+
+                if (block == null) {
+                    continue;
+                }
+
+                Block mblock = store.getChainBlockByNumber(block.getNumber());
+
+                if (mblock == null) {
+                    continue;
+                }
+
+                if (Arrays.equals(blockHash, mblock.getHash().getBytes())) {
+                    return getInternal(transactionHash, blockHash);
+                }
+            }
+
+            return Optional.empty();
+        }
+
+        // fallback with older data format
+        return super.getInMainChain(transactionHash, store);
+    }
+
+    private Optional<TransactionInfo> getInternal(byte[] transactionHash, byte[] blockHash) {
+        byte[] key = getCombinedKey(transactionHash, blockHash);
+        byte[] txInfoBytes = receiptsDS.get(key);
+        if (txInfoBytes != null) {
+            return Optional.of(new TransactionInfo(txInfoBytes));
+        }
+
+        return Optional.empty();
+    }
+
+    private static List<byte[]> parseBlockHashList(byte[] txsBytes) {
+        if (txsBytes == null || txsBytes.length == 0) {
+            return Collections.emptyList();
+        }
+
+        RLPList txHashList = (RLPList) RLP.decode2(txsBytes).get(0);
+        if (txHashList.size() == 0 || txHashList.get(0) instanceof RLPList) {
+            return Collections.emptyList();
```

Comment:
I presume than when there is an instance of RLPList, then it is the old format. Maybe we could extract now the block hashes from the old format

Reply:
yes, correct. Here we don't really need block hashes extracted from older data as they cannot be used as keys for extracting saved receipts with new format.